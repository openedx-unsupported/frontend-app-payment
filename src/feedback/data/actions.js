export const ADD_MESSAGE = 'ADD_MESSAGE';
export const REMOVE_MESSAGE = 'REMOVE_MESSAGE';

export const addMessage = (code, message, data, severity, fieldName = null) => ({
  type: ADD_MESSAGE,
  payload: {
    code,
    message,
    data,
    severity,
    fieldName,
  },
});

export const removeMessage = id => ({
  type: REMOVE_MESSAGE,
  payload: {
    id,
  },
});
