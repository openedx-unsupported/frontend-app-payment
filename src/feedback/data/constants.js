export const DANGER = 'danger';
export const SUCCESS = 'success';
export const INFO = 'info';
export const WARNING = 'warning';
