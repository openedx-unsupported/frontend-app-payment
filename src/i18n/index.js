import { messages as paragonMessages } from '@edx/paragon';
import { messages as headerMessages } from '@edx/frontend-component-header';
import { messages as footerMessages } from '@edx/frontend-component-footer';

import arMessages from './messages/ar.json';
import frMessages from './messages/fr.json';
import es419Messages from './messages/es_419.json';
import es419MessagesProtected from '../subscription/i18n-protected-messages/es_419.json';
import zhcnMessages from './messages/zh_CN.json';
import ptMessages from './messages/pt.json';
import itMessages from './messages/it.json';
import ukMessages from './messages/uk.json';
import deMessages from './messages/de.json';
import ruMessages from './messages/ru.json';
import hiMessages from './messages/hi.json';
import frCAMessages from './messages/fr_CA.json';
import deDEMessages from './messages/de_DE.json';
import itITMessages from './messages/it_IT.json';
import ptPTMessages from './messages/pt_PT.json';
// no need to import en messages-- they are in the defaultMessage field

const appMessages = {
  ar: arMessages,
  // override the protected translations
  'es-419': { ...es419Messages, ...es419MessagesProtected },
  fr: frMessages,
  'zh-cn': zhcnMessages,
  pt: ptMessages,
  it: itMessages,
  de: deMessages,
  hi: hiMessages,
  'fr-ca': frCAMessages,
  ru: ruMessages,
  uk: ukMessages,
  'de-de': deDEMessages,
  'it-it': itITMessages,
  'pt-pt': ptPTMessages,
};

export default [
  headerMessages,
  footerMessages,
  paragonMessages,
  appMessages,
];
