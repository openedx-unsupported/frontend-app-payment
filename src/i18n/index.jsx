import arMessages from './messages/ar.json';
import frMessages from './messages/fr.json';
import es419Messages from './messages/es_419.json';
import es419MessagesProtected from '../subscription/i18n-protected-messages/es_419.json';
import zhcnMessages from './messages/zh_CN.json';
import ptMessages from './messages/pt.json';
import itMessages from './messages/it.json';
import ukMessages from './messages/uk.json';
import deMessages from './messages/de.json';
import ruMessages from './messages/ru.json';
import hiMessages from './messages/hi.json';
import frCAMessages from './messages/fr_CA.json';
import deDEMessages from './messages/de_DE.json';
import itITMessages from './messages/it_IT.json';
import ptPTMessages from './messages/pt_PT.json';
import faMessages from './messages/fa.json';
import faIRMessages from './messages/fa_IR.json';
// no need to import en messages-- they are in the defaultMessage field

const messages = {
  ar: arMessages,
  // override the protected translations
  'es-419': { ...es419Messages, ...es419MessagesProtected },
  fr: frMessages,
  'zh-cn': zhcnMessages,
  pt: ptMessages,
  it: itMessages,
  de: deMessages,
  hi: hiMessages,
  'fr-ca': frCAMessages,
  ru: ruMessages,
  uk: ukMessages,
  'de-de': deDEMessages,
  'it-it': itITMessages,
  'pt-pt': ptPTMessages,
  fa: faMessages,
  'fa-ir': faIRMessages,
};

export default messages;
