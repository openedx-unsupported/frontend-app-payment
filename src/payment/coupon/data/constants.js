export const PERCENTAGE_BENEFIT = 'Percentage';
export const ABSOLUTE_BENEFIT = 'Absolute';
