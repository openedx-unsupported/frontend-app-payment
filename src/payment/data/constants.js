export const ORDER_TYPES = {
  BULK_ENROLLMENT: 'Enrollment Code',
  ENTITLEMENT: 'Course Entitlement',
  SEAT: 'Seat',
};

export const CERTIFICATE_TYPES = {
  VERIFIED: 'verified',
  CREDIT: 'credit',
};

/**
 * Payment State for async payment processing and UI Dialog Control.
 *
 * @see POLLING_PAYMENT_STATES
 *
 * @note: This enum is unique to Commerce Coordinator backend.
 */
export const PAYMENT_STATE = (((webserviceEnum = {
  // The enum as the WS Sees it.
  /**
   * Draft (Checkout) Payment
   */
  CHECKOUT: 'checkout',
  /**
   * Payment Complete
   */
  COMPLETED: 'completed',
  /**
   * Server Side Payment Failure
   */
  FAILED: 'failed',
  /**
   * Payment is Pending
   */
  PENDING: 'pending',
  /**
   * Async Processing of payment is ongoing.
   */
  PROCESSING: 'processing',
}) => ({
  ...webserviceEnum,

  // Our Additions

  /**
   * Default according to Redux initial state. (Should be an alias of an official value)
   *
   * @see PAYMENT_STATE.CHECKOUT
   */
  DEFAULT: webserviceEnum.CHECKOUT,

  /**
   * An HTTP Error has occurred between the client and server, this should not be sent over the line.
   *
   * @note **this is custom to the MFE**, thus a Symbol (which JSON usually skips in serialization)
   */
  HTTP_ERROR: Symbol('mfe-only_http_error'),
}))());

/**
 * An array of payment states that we intend to run polling against
 * @type {(string|Symbol)[]} Values from PAYMENT_STATE
 * @see PAYMENT_STATE
 */
export const POLLING_PAYMENT_STATES = [
  PAYMENT_STATE.PROCESSING,
  PAYMENT_STATE.HTTP_ERROR,
];
