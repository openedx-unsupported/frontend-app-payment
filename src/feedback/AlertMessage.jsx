import React from 'react';
import PropTypes from 'prop-types';
import { StatusAlert } from '@edx/paragon';
import { ALERT_TYPES, MESSAGE_TYPES } from './data/constants';

// Put in a message type, get an alert type.
const severityMap = {
  [MESSAGE_TYPES.DEBUG]: ALERT_TYPES.WARNING,
  [MESSAGE_TYPES.INFO]: ALERT_TYPES.INFO,
  [MESSAGE_TYPES.SUCCESS]: ALERT_TYPES.SUCCESS,
  [MESSAGE_TYPES.WARNING]: ALERT_TYPES.WARNING,
  [MESSAGE_TYPES.ERROR]: ALERT_TYPES.DANGER,
};

const AlertMessage = (props) => {
  const {
    id, messageType, userMessage, closeHandler,
  } = props;

  const onClose = () => {
    closeHandler(id);
  };

  const severity =
    messageType !== null && severityMap[messageType] !== undefined
      ? severityMap[messageType]
      : ALERT_TYPES.WARNING;

  return <StatusAlert open alertType={severity} dialog={userMessage} onClose={onClose} />;
};

AlertMessage.propTypes = {
  id: PropTypes.number.isRequired,
  messageType: PropTypes.string.isRequired,
  userMessage: PropTypes.string,
  closeHandler: PropTypes.func.isRequired,
};

AlertMessage.defaultProps = {
  userMessage: null,
};

export default AlertMessage;
