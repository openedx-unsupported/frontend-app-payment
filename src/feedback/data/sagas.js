import { put } from 'redux-saga/effects';

import { addMessage, clearMessages } from './actions';
import { MESSAGE_TYPES } from './constants';

export function* handleErrors(e, clearExistingMessages) {
  if (clearExistingMessages) {
    yield put(clearMessages());
  }

  // If this doesn't contain anything we understand, add a fallback error message
  if (e.errors === undefined && e.fieldErrors === undefined && e.messages === undefined) {
    yield put(addMessage('fallback-error', null, {}, MESSAGE_TYPES.ERROR));
  }
  if (e.errors !== undefined) {
    for (let i = 0; i < e.errors.length; i++) { // eslint-disable-line no-plusplus
      const error = e.errors[i];
      if (error.code === 'sku-error-message') {
        yield put(addMessage(error.code, error.userMessage, {}, MESSAGE_TYPES.ERROR));
      } else if (error.data === undefined && error.messageType === null) {
        yield put(addMessage('transaction-declined-message', error.userMessage, {}, MESSAGE_TYPES.ERROR));
      } else {
        yield put(addMessage(error.code, error.userMessage, error.data, error.messageType));
      }
    }
  }
  if (e.messages !== undefined) {
    for (let i = 0; i < e.messages.length; i++) { // eslint-disable-line no-plusplus
      const message = e.messages[i];
      yield put(addMessage(message.code, message.userMessage, message.data, message.messageType));
    }
  }
}

export function* handleMessages(messages, clearExistingMessages, url) {
  // If this doesn't contain anything we understand, bail.
  if (!Array.isArray(messages)) {
    return null;
  }

  if (clearExistingMessages) {
    yield put(clearMessages());
  }

  // Display an error message if one is passed through as a url parameter
  if (url && url.indexOf('error_message') > 0) {
    const errorMessage = new URLSearchParams(url).get('error_message');
    yield put(addMessage('error_message', `${errorMessage}`, {}, MESSAGE_TYPES.ERROR));
  }

  for (let i = 0; i < messages.length; i++) { // eslint-disable-line no-plusplus
    const message = messages[i];
    yield put(addMessage(message.code, message.userMessage, message.data, message.messageType));
  }
  return null;
}
