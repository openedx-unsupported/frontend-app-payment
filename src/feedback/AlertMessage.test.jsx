import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import AlertMessage from './AlertMessage';
import { MESSAGE_TYPES } from './data/constants';

describe('AlertMessage', () => {
  // The AlertList test covers most of AlertMessage testing.

  it('should handle closing', () => {
    const closeHandlerMock = jest.fn();

    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          messageType={MESSAGE_TYPES.ERROR}
          userMessage="Wondrous message!"
          closeHandler={closeHandlerMock}
        />
      </IntlProvider>
    );

    const wrapper = mount(component);
    wrapper.find('.btn').simulate('click');

    expect(closeHandlerMock).toHaveBeenCalledWith(123);
  });

  it('should default its severity when necessary', () => {
    const closeHandlerMock = jest.fn();

    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          messageType="unknown"
          userMessage="Wondrous message!"
          closeHandler={closeHandlerMock}
        />
      </IntlProvider>
    );

    const tree = renderer.create(component).toJSON();
    // The alert should have an 'alert-warning' class.  That's the default in the code.
    expect(tree).toMatchSnapshot();
  });

  it('should render a userMessage function', () => {
    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          userMessage={() => 'Wondrous message!'}
          closeHandler={jest.fn()}
        />
      </IntlProvider>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a userMessage element', () => {
    const component = (
      <IntlProvider locale="en">
        <AlertMessage
          id={123}
          userMessage={<span>Wondrous message!</span>}
          closeHandler={jest.fn()}
        />
      </IntlProvider>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
