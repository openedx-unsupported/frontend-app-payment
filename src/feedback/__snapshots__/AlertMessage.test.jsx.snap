// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertMessage should default its severity when necessary 1`] = `
<div
  className="fade alert-content alert alert-warning show"
  dialog="Wondrous message!"
  role="alert"
>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      Wondrous message!
    </div>
    <div
      className="pgn__alert-actions pgn__action-row"
    >
      <span
        className="pgn__action-row-spacer"
      />
      <button
        className="btn btn-tertiary btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </div>
  </div>
</div>
`;

exports[`AlertMessage should render a userMessage element 1`] = `
<div
  className="fade alert-content alert alert-warning show"
  dialog={
    <span
      values={Object {}}
    >
      Wondrous message!
    </span>
  }
  role="alert"
>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      <span
        values={Object {}}
      >
        Wondrous message!
      </span>
    </div>
    <div
      className="pgn__alert-actions pgn__action-row"
    >
      <span
        className="pgn__action-row-spacer"
      />
      <button
        className="btn btn-tertiary btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </div>
  </div>
</div>
`;

exports[`AlertMessage should render a userMessage function 1`] = `
<div
  className="fade alert-content alert alert-warning show"
  dialog={
    <userMessage
      values={Object {}}
    />
  }
  role="alert"
>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      Wondrous message!
    </div>
    <div
      className="pgn__alert-actions pgn__action-row"
    >
      <span
        className="pgn__action-row-spacer"
      />
      <button
        className="btn btn-tertiary btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </div>
  </div>
</div>
`;
